metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala

```diff
@@ -2,25 +2,40 @@ package scala.meta.languageserver.providers
 
 import java.nio.file.Files
 import scala.meta.languageserver.Formatter
+import scala.meta.languageserver.Configuration
 import com.typesafe.scalalogging.LazyLogging
 import langserver.messages.DocumentFormattingResult
 import langserver.types.Position
 import langserver.types.Range
 import langserver.types.TextEdit
 import org.langmeta.inputs.Input
 import org.langmeta.io.AbsolutePath
+import monix.reactive.Observable
+import monix.eval.Task
+
+class DocumentFormattingProvider(configuration: Observable[Configuration], cwd: AbsolutePath) extends LazyLogging {
+
+  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
+
+  private def formatterFromConfiguration: Task[Formatter] = configuration.lastL.map { conf =>
+    if (conf.scalafmt.enable) scalafmt
+    else Formatter.noop
+  }
+
+  private def scalafmtConfigFromConfiguration: Task[AbsolutePath] = configuration.lastL.map { conf =>
```

Comment:
This is just fine, but for general code review feedback, I would prefer to keep concepts together than are not useful separately. It feels to me like the missing concept here is that we just want a stream of format functions so we can always use the most recent.  That might be more clear if we just supplied an Observable[Formatter] or even Observable[FormattingFunction]

```scala
class DocumentFormattingProvider(configuration: Observable[Configuration], cwd: AbsolutePath) extends LazyLogging {

  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
  
  type FormatFunction = (String, String) => String
  val noFormat: FormatFunction = (str, _) => str
  val configuredFormatters: Observable[FormatFunction] = configuration.map{ conf =>
    if (conf.scalafmt.enable) {
      val fmtConf = cwd.resolve(conf.scalafmt.confPath)
      scalafmt.format(_, _, fmtConf)
    }
    else noFormat
  }
  
  def format(
      input: Input.VirtualFile
  ): Task[DocumentFormattingResult] =
    configuredFormatters.lastL.map{ fmt =>
          val fullDocumentRange = Range(
            start = Position(0, 0),
            end = Position(Int.MaxValue, Int.MaxValue)
          )
          val edits = {
            val formattedContent = fmt(input.value, input.path)
            List(TextEdit(fullDocumentRange, formattedContent))
          }
          DocumentFormattingResult(edits)
    }
}
```

Reply:
ah, I see. That's a very good point, and I actually had some doubts while making this refactor.
It's admittedly biased by the previous implementation, so I opted for the "minimum distance" change.
I'll give your feedback more thoughts later today, as I'm currently heading out right now.

Reply:
I like small step refactors as well and as long as it is an improvement (which I say this is) I like to pull it in, we can always do another round of refactorings later.

Reply:
@ShaneDelmore `Formatter` is your `FormatFunction` 😜 I personally prefer methods + named parameters over higher-order functions on tuples. The former are easier to read, debug, use in the IDE, ...

Reply:
I tend to prefer streams with functions applied to them (especially over callbacks), but I will keep try to keep your preferences in mind.

Reply:
@olafurpg FWIW the most important thing to me usually is dealing in whole concepts when possible, position vs (line, column) and here my preferences is not a formatter function in general, but to get an observable of formatter, vs an unconfigured formatter and a format conf that need to be assembled before use when all anyone really wants is the thing you call to format the text.